### ISSUE TITLE
Graphics + text split screen modes

### ISSUE THREAD

**Issue by dansanderson** [MEGA65/mega65-rom-public#12]
Bitmap graphics modes accessible with BASIC commands are great, but pose a disadvantage for beginners when graphics and text cannot appear on the same display, because it is difficult to experiment with drawing commands in immediate mode. Vintage microcomputers including the Commodore 128, 16, and Plus/4 accommodate this use case (and others) with a spilt screen display mode.

On the 128, the BASIC 7 command `GRAPHIC <n>[,<clear>[,<line>]]` selects between 40-col text, standard bitmap, split standard bitmap + 40-col text, multi-color bitmap, split multi-color bitmap + 40-col text, and 80-col text display modes. `clear`=1 would clear the bitmap screen (default 0). `line` would set the split point at a given text row (default 20, leaving five text lines). Only one vertical split is possible and the graphics mode is always on top. There's also a separate `SCNCLR` for clearing one of the screens without changing modes. `GRAPHIC CLR` and BASIC errors revert to 40-col text mode.

![CleanShot 2022-06-03 at 11 31 22](https://user-images.githubusercontent.com/408973/171925033-f6abc884-d157-4c7f-bf92-6e137349ca71.png)

BASIC 65 has a more sophisticated screen system managed by the `SCREEN` commands, with up to four bitmap graphics screens in any of the supported resolutions and color bit depths. It does not yet support a graphics + text split screen mode or the BASIC 7 `GRAPHIC <n>` commands. (`GRAPHIC CLR` and `SCNCLR` are implemented.)

## Proposal

Supplement the M65 screen system with a graphics+text split capability. Implement a similar `GRAPHIC` command to control it. The bitmap region would be the current bitmap viewscreen. If there isn't a current bitmap viewscreen, screen 0 would be allocated with sensible defaults similar to the `SCREEN` command. I imagine that horizontal screen resolution cannot be changed mid-frame, so a M65 split screen mode would use the text mode (40/80 col) that corresponds with the current viewscreen.

`GRAPHIC <n>[,<clear>[,<line>]]` would retain its C128 syntax, with a different mode list: 1) 40 col text, 2) 80 col text, 3) split 320-wide bitmap + 40 text, 4) split 640-wide bitmap + 80 text. Modes 3 and 4 would defer to the active bitmap viewscreen resolution if there is one.

As with bitmap screen modes, `GRAPHIC CLR`, `SCREEN CLOSE`, and BASIC errors would restore text mode.

If `line` is provided, `GRAPHIC` also clears the text screen and relocates the next READY prompt to the first visible text line. This matches the C128 behavior. Note that the C128 neither clears the text screen nor relocates the cursor if `line` is not provided. The proposed solution is to mimic the C128 behavior for consistency except where it doesn't make sense.

## Alternatives

I'm hoping that this can be implemented with a simple raster interrupt and manipulation of the existing VIC-IV screen modes, but I have not tried this myself yet. Perhaps a beginner's tool can implement this as a small ML program that one can load from disk or add to their own programs, with no BASIC extension necessary. Given the way VICE flickers in this mode I'm guessing this is how it is actually implemented on the C128. I would expect other raster interrupt routines (VSYNC?) and RUN/STOP-RESTORE to cancel split screen mode.

We could dispense with multiple modes for `GRAPHIC` and just use the current bitmap viewscreen or the current text mode (in that order) to determine the mode. A user that wants a specific bitmap resolution and bit depth can call `SCREEN` before calling `GRAPHIC`. I like how this turns `GRAPHIC` without arguments into a simple intuitively named toggle for quickly jumping in and out of split-screen graphics mode.

I assumed that 640 bitmap and 40 col text, or 320 bitmap and 80 col text, cannot coexist on the same display. If I'm wrong, maybe more options (perhaps via numbered `GRAPHIC` modes) make sense.

I never liked how the cursor is allowed behind the bitmap display. Perhaps a `WINDOW` could be established automatically. Maybe Home-Home should also exit split screen mode as well as exit the WINDOW.

**Comment by dansanderson** [comment]
P.S. I'm willing to do some poking around on possible implementation paths. I'd like feedback and confirmation from the leadership team before investing in a full implementation, as it'll probably consume some space that some may consider unjustified.

**Comment by gurcei** [comment]
I'll admit I'm in the camp of folks that wished we had a graphics+text split screen mode. It's a feature on the C128 that I looked on with envy in my C64 days, and when I thought I could finally enjoy it on the MEGA65, it was gone :-(

I agree it will make it easier for users to explore and experiment with graphics commands and ease their learning curve there.

Anyways, will defer to Bit Shifter on the feasibility and available space side of things.

**Comment by Edilbert** [comment]
This would be indeed a useful screen mode.
Also having a dual screen mode (e.g. HDMI text and VGA graphics) would be nice.

I think, this can be done by extending the capabilities of the VIC-IV with VHDL code in the core.
However I never programmed VHDL myself and cannot tell, how much effort that would be.

Trying to implement this in ROM is probably not a good solution.
Maybe, it is technically possible, but switching the screen mode on every frame at a special raster line
between graphics and text would require a complex re-programming of the VIC chip affecting probably
30 registers (my estimate). The VIC-IV has so many registers and hot mode cannot be used in this case, I assume.

To summarize: I think it's a good idea, but recommend to implement it in the core.

I would like to read Paul's comment on this topic, who has a better knowledge of the VIC-IV.

**Comment by gurcei** [comment]
Oh, the thought of a core/vhdl solution made me think of matrix-mode. Perhaps on a unique key-press we could re-purpose that mode to appear as a hard split-screen rather than a transparent overlay, and instead of routing keypresses to the uart monitor, just let them continue to be listened to ny BASIC/ROM?

**Comment by lgblgblgb** [comment]
Btw it almost looks like as a need for a kind-of Amiga feature, ie a hardware list that given rasterline(s) hit by VIC-IV should trigger to write a certain register(s) with a certain value(s). cooper-list, if I remember correctly maybe? Or kind of that only simplified form? So basically almost like a "hand written" raster IRQ-handler just implemented in hardware. It can be used for many other things as well though, not only for screen split. Surely, at this point it's already a mega65-core wish, not a ROM feature, or at least not enterally, sure the ROM can still use this imagine feature than, as any other program too.

**Comment by dansanderson** [comment]
I'm closing this as will-not-fix. I think the better choice for this is our tentative plan to replace the BASIC bitmap graphics system with one based on FCM character graphics, which will allow easy combining of text and graphics, in combination with WINDOW. It'll be a large project, but it's more likely than a raster split between character and bitplane modes (without a core change to support it).
